### ISSUE TITLE
2.6.1 0010_populate_translation_project_field.py migration is failing

### ISSUE THREAD

**Issue by dwaynebailey** [translate/pootle#4101]
Migration issue reported on [``pootle-devel``](http://markmail.org/message/uieojvgity3puid5)

Summary is that a cleanup migration seems to be failing.

Migration code: https://github.com/translate/pootle/blob/stable/2.6.1/pootle/apps/pootle_store/migrations/0010_populate_translation_project_field.py#L9-L14

As an aside. Can we actually document in the migration what the migration is hoping to achieve.

```pytb 

> pootle_store:0010_populate_translation_project_field
 - Migration 'pootle_store:0010_populate_translation_project_field' is
marked for no-dry-run.
 ! Error found during real run of migration! Aborting.

 ! Since you have a database that does not support running
 ! schema-altering statements in transactions, we have had
 ! to leave it in an interim state between migrations.

! You *might* be able to recover with:   (migration cannot be dry-run;
cannot discover commands)
 ! The South developers regret this has happened, and would
 ! like to gently persuade you to consider a slightly
 ! easier-to-deal-with DBMS (one that supports DDL transactions)
 ! NOTE: The error which caused the migration to fail is further up.
Error in migration: pootle_store:0010_populate_translation_project_field
```
```pytb
Traceback (most recent call last):
  File "/var/www/pootle/env/bin/pootle", line 11, in <module>
    sys.exit(main())
  File "/var/www/pootle/env/lib/python2.7/site-packages/pootle/runner.py", line 171, in main
    django_settings_module='pootle.settings')
  File "/var/www/pootle/env/lib/python2.7/site-packages/pootle/runner.py", line 150, in run_app
    management.execute_from_command_line([runner_name] + remainder)
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/base.py", line 415, in handle
    return self.handle_noargs(**options)
  File "/var/www/pootle/env/lib/python2.7/site-packages/pootle/apps/pootle_app/management/commands/setup.py", line 77, in handle_noargs
    call_command('migrate', interactive=False)
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 159, in call_command
    return klass.execute(*args, **defaults)
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/management/commands/migrate.py", line 111, in handle
    ignore_ghosts = ignore_ghosts,
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/__init__.py", line 220, in migrate_app
    success = migrator.migrate_many(target, workplan, database)
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/migrators.py", line 256, in migrate_many
    result = migrator.__class__.migrate_many(migrator, target, migrations, database)
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/migrators.py", line 331, in migrate_many
    result = self.migrate(migration, database)
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/migrators.py", line 133, in migrate
    result = self.run(migration, database)
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/migrators.py", line 114, in run
    return self.run_migration(migration, database)
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/migrators.py", line 84, in run_migration
    migration_function()
  File "/var/www/pootle/env/lib/python2.7/site-packages/south/migration/migrators.py", line 60, in <lambda>
    return (lambda: direction(orm))
  File "/var/www/pootle/env/lib/python2.7/site-packages/pootle/apps/pootle_store/migrations/0010_populate_translation_project_field.py", line 13, in forwards
    bs.translation_project = bs.unit.store.translation_project
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/db/models/fields/related.py", line 320, in __get__
    rel_obj = qs.get()
  File "/var/www/pootle/env/lib/python2.7/site-packages/django/db/models/query.py", line 310, in get
    self.model._meta.object_name)
pootle_store.models.DoesNotExist: Unit matching query does not exist.
```

**Comment by phlax** [comment]
looks like orphaned submissions to me

**Comment by unho** [comment]
Yes, only explanation is that that install has suggestions with no unit. I will just add a quick `try` block there so in case of exception we just remove the suggestion:

```python
    def forwards(self, orm):
        broken_suggestions = orm['pootle_store.Suggestion'].objects.filter(translation_project__isnull=True)

        for bs in broken_suggestions:
            try:
                bs.translation_project = bs.unit.store.translation_project
            except Exception:
                # There is something wrong with the suggestion, e.g. the unit
                # it references doesn't exist, so just delete the suggestion.
                bs.delete()
            else:
                bs.save()
```

@phlax Does this looks ok to you?

**Comment by phlax** [comment]
@unho just putting a PR for this now

**Comment by unho** [comment]
@phlax I am thinking that just using `Exception` might hide other problems, so perhaps better use https://docs.djangoproject.com/en/1.6/ref/exceptions/#django.core.exceptions.DoesNotExist

**Comment by phlax** [comment]
@unho - in this case i think we should check if the suggestion has a unit rather than try/except

But in general `except Exception` is very bad practice

**Comment by phlax** [comment]
actually @unho - my bad - we do need to use try/except - ill update

**Comment by unho** [comment]
Yes, I was thinking about the scenario where the suggestion has a unit PK there, but no unit with that PK exists. I can't actually remember what Django returns in case you try to get the suggestion unit in that scenario, but my guess is DoesNotExist.

**Comment by dwaynebailey** [comment]
To test this in the migration replace:

``pip install --upgrade "Pootle>=2.6,<2.7"``
with
``pip install --upgrade git://github.com/translate/pootle@2.6.x#egg=pootle``

That will pull in the latest 2.6 branch with the changes we'd like you to test.

**Comment by dwaynebailey** [comment]
@bgkriegel could you follow the [instructions above](https://github.com/translate/pootle/issues/4101#issuecomment-143191461) and give us feedback on your results.  We'll deal with any other migration issues in other bugs if we hit them.  But we'd like to know that you get a clean migration before we release a ``2.6.2`` version.

**Comment by bgkriegel** [comment]
@dwaynebailey, it still doesn't work.
I can attach full trace if needed, but diff between new an old is:

(env)root@host:/var/www/pootle# diff newtrace oldtrace 
2,3c2,3
<   File "/var/www/pootle/env/bin/pootle", line 9, in <module>
<     load_entry_point('Pootle==2.6.1', 'console_scripts', 'pootle')()
---
>   File "/var/www/pootle/env/bin/pootle", line 11, in <module>
>     sys.exit(main())
40,41c40,41
<   File "/var/www/pootle/env/lib/python2.7/site-packages/pootle/apps/pootle_store/migrations/0010_populate_translation_project_field.py", line 20, in forwards
<     bs.translation_project = unit.store.translation_project
---
>   File "/var/www/pootle/env/lib/python2.7/site-packages/pootle/apps/pootle_store/migrations/0010_populate_translation_project_field.py", line 13, in forwards
>     bs.translation_project = bs.unit.store.translation_project
46c46
< pootle_store.models.DoesNotExist: Store matching query does not exist.
---
> pootle_store.models.DoesNotExist: Unit matching query does not exist.

**Comment by dwaynebailey** [comment]
@phlax I wonder if this is a different case?  Now we have a missing Store.

**Comment by unho** [comment]
@bgkriegel Thank you very much for the time you are investing in tests. Can you please try again using the same [instructions above](https://github.com/translate/pootle/issues/4101#issuecomment-143191461) and give us feedback on your results.

**Comment by bgkriegel** [comment]
@unho Good news!! This time it went smoothly :)
Only two warnings after 144 minutes, 
checked twice, just in case.

<pre>
 > accounts:0002_import_profile_data
Importing old profile data
Importing alt_src_langs data. If you have a lot of users, this can take a while.
Renaming u'pootle_app_pootleprofile_alt_src_langs' to u'accounts_user_alt_src_langs'
Importing old profile data
2015-09-25 21:28:48,346 WARNING /var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py:124: Warning: Column 'unit_rows' cannot be null
  return self.cursor.execute(query, args)

...

 > socialaccount:0012_auto__chg_field_socialtoken_token_secret
 - Loading initial data for socialaccount.
Installed 0 object(s) from 0 fixture(s)
2015-09-25 23:19:49,669 INFO Detected new Pootle version: 26100.
2015-09-25 23:19:49,670 INFO Running the upgrade machinery...
2015-09-25 23:19:49,709 INFO Succesfully migrated old build versions to new PootleConfig.
2015-09-25 23:19:49,761 WARNING /var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/managers.py:31: DeprecationWarning: The "depth" keyword argument has been deprecated.
Use related field names instead.
  .select_related(depth=1)

2015-09-25 23:19:49,825 INFO Database now at Pootle build 25201
2015-09-25 23:19:49,900 INFO Database now at Pootle build 25203
2015-09-25 23:19:49,901 INFO About to synchronize latest submission data.
2015-09-25 23:22:19,497 INFO Succesfully synchronized latest submission data.
2015-09-25 23:22:19,569 INFO Database now at Pootle build 25205
2015-09-25 23:22:19,603 INFO Database now at Toolkit build 12008
2015-09-25 23:22:19,636 INFO Database now at Pootle build 25206
2015-09-25 23:22:19,638 INFO Updating 'view' permission
2015-09-25 23:22:19,640 INFO Creating 'hide' permission
2015-09-25 23:22:19,711 INFO Database now at Pootle build 25999
2015-09-25 23:22:19,797 INFO Done.
2015-09-25 23:22:19,798 WARNING


    Warning: Pootle 2.6.1 is an interim release (a migration step to Pootle
             2.7.0). Do not use Pootle 2.6.1 for any deployment.


</pre>

I'll try to continue to test 2.7.2
Many thanks

**Comment by unho** [comment]
Yes, please upgrade to 2.7.2.

Closing this. We will release 2.6.2. Soon.

**Comment by dwaynebailey** [comment]
Fixed in ba40a563050951d76e3945b63a462c4cb6e26cae
